### PR TITLE
Add HEAD fix to gitea doctor (#21352)

### PR DESCRIPTION
**Pull request by ghost**
Backport #21352

Due to a bug in presumably an older version of Gitea, multiple of my repositories still have their HEADs pointing to a `master` branch while the default branch on the UI is listed as `main`. This adds a `gitea doctor` command that will fix all of the HEAD references for repos when they're not synchronized with the default branch in the DB.

This will help with cloning to ensure that git automatically checks out the right branch, instead of a nonexistent one.

Note: I'm not sure if I actually need to do more other than add a file here. Will try testing this out on my server soon.

Co-authored-by: zeripath <art27@cantab.net>
Co-authored-by: wxiaoguang <wxiaoguang@gmail.com>

### CODE REVIEW

**Review by KN4CK3R — APPROVED** [review]

File: modules/doctor/heads.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+		if headErr != nil {
+			numHeadsBroken++
+		}
+		if defaultBranchErr != nil {
+			numDefaultBranchesBroken++
+		}
+
+		// if default branch is broken, let the user fix that in the UI
+		if defaultBranchErr != nil {
```
Comment (delvh) [inline modules/doctor/heads.go:41]:
```suggestion
		
		// if default branch is broken, let the user fix that in the UI
		if defaultBranchErr != nil {
			numDefaultBranchesBroken++
```

File: modules/doctor/heads.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+			logger.Info("All %d repos have their HEADs in the correct state")
+		} else {
+			if numHeadsBroken == 0 && numDefaultBranchesBroken != 0 {
+				logger.Critical("Default branches are broken for %d/%d repos", numDefaultBranchesBroken, numRepos)
+			} else if numHeadsBroken != 0 && numDefaultBranchesBroken == 0 {
+				logger.Warn("HEADs are broken for %d/%d repos", numHeadsBroken, numRepos)
+			} else {
+				logger.Critical("Out of %d repos, HEADS are broken for %d and default branches are broken for %d", numRepos, numHeadsBroken, numDefaultBranchesBroken)
+			}
+		}
```
Comment (delvh) [inline modules/doctor/heads.go:77]:
```suggestion
		logger.Info("Out of %d repos, HEADs for %d are now fixed and HEADS for %d are still broken", numRepos, numReposUpdated, numDefaultBranchesBroken+numHeadsBroken-numReposUpdated)
		return nil
	}
	
	if numHeadsBroken == 0 && numDefaultBranchesBroken == 0 {
		logger.Info("All %d repos have their HEADs in the correct state")
		return nil
	}
	
	if numHeadsBroken == 0 && numDefaultBranchesBroken != 0 {
		logger.Critical("Default branches are broken for %d/%d repos", numDefaultBranchesBroken, numRepos)
	} else if numHeadsBroken != 0 && numDefaultBranchesBroken == 0 {
		logger.Warn("HEADs are broken for %d/%d repos", numHeadsBroken, numRepos)
	} else {
		logger.Critical("Out of %d repos, HEADS are broken for %d and default branches are broken for %d", numRepos, numHeadsBroken, numDefaultBranchesBroken)
	}
```

**Review by delvh — APPROVED** [review]
Oh wait, I just noticed that this is a backport.
Ignore the stylistic things then :sweat:

**Comment by lunny** [comment]
make L-G-T-M work

**Pull request merged**